src/MusicWheel.cpp

```diff
@@ -673,38 +673,67 @@ void MusicWheel::BuildWheelItemDatas( std::vector<MusicWheelItemData *> &arrayWh
 						break;
 				}
 			}
-
 			// make WheelItemDatas with sections
 			RString sLastSection = "";
 			int iSectionColorIndex = 0;
-			for( unsigned i=0; i< arraySongs.size(); i++ )
+
+			// If the sort order is Preferred handle it differently because we already know the sections
+			if( so == SORT_PREFERRED )
 			{
-				Song* pSong = arraySongs[i];
 				if( bUseSections )
 				{
-					RString sThisSection = SongUtil::GetSectionNameFromSongAndSort( pSong, so );
-
-					if( sThisSection != sLastSection )
+					// Get all section names
+					std::vector<RString> vsSectionNames = SONGMAN->GetPreferredSortSectionNames();
+					for( unsigned i=0; i<vsSectionNames.size(); i++ )
 					{
-						int iSectionCount = 0;
-						// Count songs in this section
-						unsigned j;
-						for( j=i; j < arraySongs.size(); j++ )
+						
+						// Get all songs in this section
+						std::vector<Song*> vsSongsInSection = SONGMAN->GetPreferredSortSongsBySectionName( vsSectionNames[i] );

```

Comment:
This is technically inefficient.

We have the external for loop that loops over section names, and then we have `GetPreferredSortSongsBySectionName` called for each section name which will loop over `m_vPreferredSongSort` for each. That's O(n^2) while ot might be better to just build the mapping from sectionname -> songs first and then key into it with section names, which is O(n).

Reply:
Alrighty just created the mapping. Lmk if that's good